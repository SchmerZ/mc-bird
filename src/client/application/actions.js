import createAction from "redux-actions/es/createAction";

export const navigateTo = createAction('APPLICATION_NAVIGATE_TO_PAGE');
export const notify = createAction('APPLICATION_NOTIFICATION_NOTIFY');
export const messageClose = createAction('APPLICATION_NOTIFICATION_MESSAGE_CLOSE');
