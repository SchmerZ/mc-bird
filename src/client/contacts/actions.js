import {createAction} from 'redux-actions'
import {createRoutine} from 'redux-saga-routines'

export const init = createAction('CONTACTS_INIT');
export const fetch = createRoutine('CONTACTS_FETCH');

export const prevPage = createAction('CONTACTS_PREV_PAGE');
export const nextPage = createAction('CONTACTS_NEXT_PAGE');

export const navigateToConversation = createAction('CONTACTS_NAVIGATE_TO_CONVERSATION');
export const messageAdd = createAction('CONTACTS_MESSAGE_ADD');
