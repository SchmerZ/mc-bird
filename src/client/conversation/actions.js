import {createAction} from 'redux-actions'
import {createRoutine} from 'redux-saga-routines'

export const init = createAction('CONVERSATION_INIT');
export const leave = createAction('CONVERSATION_LEAVE');
export const fetch = createRoutine('CONVERSATION_FETCH');

export const changeTypedText = createAction('CONVERSATION_TYPED_TEXT_CHANGE');
export const send = createRoutine('CONVERSATION_SEND_MESSAGE');

export const typedTextError = createAction('CONVERSATION_TYPED_TEXT_ERROR');
export const backToContacts = createAction('CONVERSATION_BACK_TO_CONTACTS');

export const messageAdd = createAction('CONVERSATION_MESSAGE_ADD');
