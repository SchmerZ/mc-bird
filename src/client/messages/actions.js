import {createAction} from 'redux-actions'
import {createRoutine} from 'redux-saga-routines'

export const init = createAction('MESSAGES_LIST_INIT');
export const fetch = createRoutine('MESSAGES_LIST_FETCH');
