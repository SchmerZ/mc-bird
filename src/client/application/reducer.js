import {handleActions} from 'redux-actions'
import * as A from './actions'

const initialState = {
  messages: [],
};

const handlers = {
  [A.notify]: (state, {payload: {message, type}}) => {
    const nextMessages = [...state.messages, {message, type}];

    return {
      ...state,
      messages: nextMessages,
    }
  },
  [A.messageClose]: (state) => {
    const [, ...restMessages] = state.messages;

    return {
      ...state,
      messages: restMessages,
    }
  },
};

export default handleActions(handlers, initialState);
