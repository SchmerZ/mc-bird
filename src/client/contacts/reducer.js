import {handleActions} from 'redux-actions'
import * as A from './actions'

const initialState = {
  fetching: true,
  fetchingOffset: 0,
  fetchingFailed: false,

  contacts: {},
  totalCount: 0,
  offset: 0,
  limit: 10,
};

const handlers = {
  [A.init]: (state) => ({
    ...state,
    contacts: {},
  }),

  [A.fetch]: (state, {payload}) => ({
    ...state,
    fetchingOffset: payload.offset,
  }),
  [A.fetch.request]: (state) => ({
    ...state,
    fetching: true,
  }),
  [A.fetch.success]: (state, {payload}) => {
    const {items, totalCount, offset} = payload;

    const hash = items.reduce((memo, curr) => ({
      ...memo,
      [curr.msisdn]: curr
    }), {});

    return {
      ...state,
      contacts: hash,
      totalCount,
      offset,
      fetching: false,
      fetchingFailed: false,
    }
  },
  [A.fetch.failure]: (state) => ({
    ...state,
    fetching: false,
    fetchingFailed: true,
  }),

  [A.nextPage]: (state) => ({
    ...state,
    fetchingOffset: Math.min(state.offset + state.limit, state.totalCount),
  }),
  [A.prevPage]: (state) => ({
    ...state,
    fetchingOffset: Math.max(state.offset - state.limit, 0),
  }),

  [A.messageAdd]: (state, {payload}) => {
    const {msisdn} = payload;
    const current = state.contacts[msisdn];

    return {
      ...state,
      contacts: {
        ...state.contacts,
        [msisdn]: {
          ...current,
          messages: {
            ...current.messages,
            totalCount: current.messages.totalCount + 1,
          }
        }
      }
    }
  }
};

export default handleActions(handlers, initialState);
