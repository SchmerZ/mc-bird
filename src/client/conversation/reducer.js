import {handleActions} from 'redux-actions'
import {createSelector} from 'reselect'

import moment from 'moment'

import * as A from './actions'

const initialState = {
  fetching: true,
  fetchingFailed: false,
  sending: false,

  contact: null,
  msisdn: null,
  messages: [],
  typedText: '',
  error: null,
};

const handlers = {
  [A.init]: (state, {payload}) => ({
    ...state,
    msisdn: payload.match.params.msisdn,
    contact: null,
    messages: [],
    error: null,
  }),

  [A.fetch.request]: (state) => ({
    ...state,
    fetching: true,
  }),
  [A.fetch.success]: (state, {payload}) => {
    const {items, contact} = payload;

    return {
      ...state,
      contact,
      messages: items,
      fetching: false,
      fetchingFailed: false,
    }
  },
  [A.fetch.failure]: (state) => ({
    ...state,
    fetching: false,
    fetchingFailed: true,
  }),

  [A.changeTypedText]: (state, {payload: value}) => ({
    ...state,
    typedText: value,
  }),

  [A.typedTextError]: (state, {payload: {error}}) => ({
    ...state,
    error,
  }),

  [A.send.request]: (state) => ({
    ...state,
    sending: true,
  }),
  [A.send.success]: (state, {payload: sentMessage}) => {
    return {
      ...state,
      messages: [sentMessage, ...state.messages],
      typedText: '',
      sending: false,
    }
  },
  [A.send.failure]: (state) => ({
    ...state,
    sending: false,
  }),

  // [A.messageAdd]: (state, {payload}) => {
  //   const {msisdn} = payload;
  //   const current = state.contacts[msisdn];
  //
  //   return {
  //     ...state,
  //     contacts: {
  //       ...state.contacts,
  //       [msisdn]: {
  //         ...current,
  //         messages: {
  //           ...current.messages,
  //           totalCount: current.messages.totalCount + 1,
  //         }
  //       }
  //     }
  //   }
  // }
};

const compare = (a, b) => {
  const ma = moment(a.createdDatetime);
  const mb = moment(b.createdDatetime);

  return ma.diff(mb);
};

const itemsInReverseOrderSelector = createSelector(
  state => state.conversation.messages,
  messages => messages.sort(compare),
);

export const SELECTORS = {
  getItemsInReverseOrder: itemsInReverseOrderSelector,
};

export default handleActions(handlers, initialState);
