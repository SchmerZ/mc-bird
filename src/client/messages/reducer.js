import {handleActions} from 'redux-actions'
import * as A from './actions'

import messageType from '../constants/message-type'

const initialState = {
  fetching: true,
  fetchingOffset: 0,
  fetchingFailed: false,

  items: [],
  totalCount: 0,
  offset: 0,
  limit: 10,
};

const handlers = {
  [A.init]: (state) => ({
    ...state,
    items: [],
  }),

  [A.fetch]: (state, {payload}) => ({
    ...state,
    fetchingOffset: payload.offset,
  }),
  [A.fetch.request]: (state) => ({
    ...state,
    fetching: true,
  }),
  [A.fetch.success]: (state, {payload}) => {
    const {items, totalCount, offset} = payload;

    return {
      ...state,
      items,
      totalCount,
      offset,
      fetching: false,
      fetchingFailed: false,
    }
  },
  [A.fetch.failure]: (state, {payload}) => ({
    ...state,
    fetching: false,
    fetchingFailed: true,
  }),

  [A.nextPage]: (state) => ({
    ...state,
    fetchingOffset: Math.min(state.offset + state.limit, state.totalCount),
  }),
  [A.prevPage]: (state) => ({
    ...state,
    fetchingOffset: Math.max(state.offset - state.limit, 0),
  }),

  [A.messageAdd]: (state, {payload}) => {
    const {message} = payload;
    const newMessage = {
      ...message,
      direction: messageType.received,
      recipients: {
        items: [{
          originator: null,
          recipient: message.recipient,
          status: "income"
        }]
      }
    };

    return {
      ...state,
      items: [newMessage, ...state.items.slice(0, state.items.length - 1)],
      totalCount: state.totalCount + 1,
    }
  }
};

export default handleActions(handlers, initialState);
