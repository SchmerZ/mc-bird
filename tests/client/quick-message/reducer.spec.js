import React from 'react';

import reducer from '../../../src/client/quick-message/reducer';
import * as A from '../../../src/client/quick-message/actions';

describe('Quick message form', () => {
  describe('Reducer', () => {
    it('should have initial state', () => {
      const nextState = reducer(undefined, {});
      expect(nextState).toMatchSnapshot();
    });

    it('should update state on recipient value change', () => {
      const nextState = reducer(undefined, A.changeRecipient('some new value'));
      expect(nextState).toMatchSnapshot();
    });

    it('should update sending flag on send routine action', () => {
      let nextState = reducer({sending: false}, A.send.request());
      expect(nextState.sending).toEqual(true);

      nextState = reducer({sending: true}, A.send.success());
      expect(nextState.sending).toEqual(false);

      nextState = reducer({sending: true}, A.send.failure());
      expect(nextState.sending).toEqual(false);
    });
  });
});
