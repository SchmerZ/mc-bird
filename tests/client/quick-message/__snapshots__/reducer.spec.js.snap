// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quick message form Reducer should have initial state 1`] = `
Object {
  "errors": Object {},
  "messageText": "",
  "recipient": "",
  "sending": false,
}
`;

exports[`Quick message form Reducer should update state on recipient value change 1`] = `
Object {
  "errors": Object {},
  "messageText": "",
  "recipient": "some new value",
  "sending": false,
}
`;
